### DISCUSSION TITLE
Input port callback not working

### DISCUSSION THREAD

**Discussion by gary-1959** [mido/mido#602] [Q&A] [answered]
I can't get an input port to fire a callback, and I can't see how this is being supported in the code. Is it supposed to? The equivalent rtmidi code works OK.

This does not work on Windows 10, Linux 24.04, Python 13 or Python 12:

```
import mido
import time

innames = mido.get_input_names()

def input_callback():
    print("MESSAGE RECEIVED")

port_in = None
for n in innames:
    if n.startswith("USB"):
        port_in = mido.open_input(n, callback=input_callback)


if port_in != None:
    while True:
        time.sleep(1)
```

**Comment by r1yk** [comment, marked as answer]
It looks like your callback function probably needs to be updated to take a single argument for the message it will receive each time it's called. From the docstring for `mido.open_input`:

```
callback=None
    A callback function to be called when a new message arrives.
    The function should take one argument (the message).
    Raises IOError if not supported by the backend.
```
You could quickly test with something like:
 ```python
def input_callback(msg):
    print("MESSAGE RECEIVED", msg)
 ```

**Comment by gary-1959** [comment]
Thanks for your help
